File: mobile/src/services/labelMetadata.js
/**
 * Label Metadata Service
 * ──────────────────────
 * JavaScript port of ai-service/app/model/label_metadata.py.
 *
 * Parses flat model labels (e.g. "Tomato___Late_blight" or
 * "Tomato - Late blight") into structured metadata objects with
 * crop name, disease name, category, and healthy status.
 *
 * This runs entirely on-device so the mobile app can produce the
 * same label structure as the server without network access.
 */

const HEALTHY_TERMS = new Set(['healthy', 'normal', 'no disease', 'none']);

/**
 * Replace underscores and hyphens with spaces and title-case the result.
 * @param {string} value
 * @returns {string}
 */
function humanize(value) {
  const cleaned = value
    .replace(/[_-]+/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();

  if (!cleaned) return 'Unknown';

  return cleaned
    .split(' ')
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
    .join(' ');
}

/**
 * Parse a model label into structured metadata.
 *
 * Supports common dataset label formats:
 * - PlantVillage:   "Tomato___Late_blight"
 * - Display labels: "Tomato - Late blight"
 * - Double under:   "Tomato__Late_blight"
 * - Generic:        "Rust", "Healthy"
 *
 * @param {string} label - Raw model output label.
 * @returns {{
 *   rawLabel: string,
 *   cropName: string,
 *   diseaseName: string,
 *   displayName: string,
 *   isHealthy: boolean,
 *   category: string,
 * }}
 */
export function parseLabel(label) {
  const rawLabel = label ? label.trim() : 'Unknown';

  let cropPart;
  let diseasePart;

  if (rawLabel.includes('___')) {
    [cropPart, diseasePart] = rawLabel.split('___', 2);
  } else if (rawLabel.includes(' - ')) {
    [cropPart, diseasePart] = rawLabel.split(' - ', 2);
  } else if (rawLabel.includes('__')) {
    [cropPart, diseasePart] = rawLabel.split('__', 2);
  } else {
    cropPart = 'Unknown';
    diseasePart = rawLabel;
  }

  const cropName = humanize(cropPart);
  let diseaseName = humanize(diseasePart);
  const normalizedDisease = diseaseName.toLowerCase();

  const isHealthy =
    HEALTHY_TERMS.has(normalizedDisease) ||
    normalizedDisease.includes('healthy');

  if (isHealthy) {
    diseaseName = 'Healthy';
  }

  let category;
  if (isHealthy) {
    category = 'healthy';
  } else if (
    ['blight', 'rot', 'mold', 'mildew', 'scab'].some((t) =>
      normalizedDisease.includes(t),
    )
  ) {
    category = 'fungal';
  } else if (
    ['virus', 'mosaic', 'curl'].some((t) => normalizedDisease.includes(t))
  ) {
    category = 'viral';
  } else if (
    ['spot', 'speck', 'bacterial'].some((t) => normalizedDisease.includes(t))
  ) {
    category = 'bacterial';
  } else if (normalizedDisease.includes('rust')) {
    category = 'fungal';
  } else {
    category = 'unknown';
  }

  const displayName =
    cropName === 'Unknown'
      ? diseaseName
      : `${cropName} - ${diseaseName}`;

  return {
    rawLabel,
    cropName,
    diseaseName,
    displayName,
    isHealthy,
    category,
  };
}

File: ai-service/app/model/label_metadata.py
from dataclasses import asdict, dataclass
import re


HEALTHY_TERMS = {"healthy", "normal", "no disease", "none"}


@dataclass(frozen=True)
class LabelMetadata:
    raw_label: str
    crop_name: str
    disease_name: str
    display_name: str
    is_healthy: bool
    category: str

    def to_dict(self) -> dict:
        return asdict(self)


def _humanize(value: str) -> str:
    cleaned = re.sub(r"[_-]+", " ", value).strip()
    cleaned = re.sub(r"\s+", " ", cleaned)
    return cleaned.title() if cleaned else "Unknown"


def parse_label(label: str) -> LabelMetadata:
    """Parse flat model labels into crop and disease metadata.

    Supports common dataset label formats:
    - PlantVillage: Tomato___Late_blight
    - Display labels: Tomato - Late blight
    - Generic single labels: Rust, Healthy
    """

    raw_label = label.strip() if label else "Unknown"

    if "___" in raw_label:
        crop_part, disease_part = raw_label.split("___", 1)
    elif " - " in raw_label:
        crop_part, disease_part = raw_label.split(" - ", 1)
    elif "__" in raw_label:
        crop_part, disease_part = raw_label.split("__", 1)
    else:
        crop_part, disease_part = "Unknown", raw_label

    crop_name = _humanize(crop_part)
    disease_name = _humanize(disease_part)
    normalized_disease = disease_name.lower()
    is_healthy = normalized_disease in HEALTHY_TERMS or "healthy" in normalized_disease
    disease_name = "Healthy" if is_healthy else disease_name

    if is_healthy:
        category = "healthy"
    elif any(term in normalized_disease for term in ("blight", "rot", "mold", "mildew", "scab")):
        category = "fungal"
    elif any(term in normalized_disease for term in ("virus", "mosaic", "curl")):
        category = "viral"
    elif any(term in normalized_disease for term in ("spot", "speck", "bacterial")):
        category = "bacterial"
    elif "rust" in normalized_disease:
        category = "fungal"
    else:
        category = "unknown"

    display_name = disease_name if crop_name == "Unknown" else f"{crop_name} - {disease_name}"

    return LabelMetadata(
        raw_label=raw_label,
        crop_name=crop_name,
        disease_name=disease_name,
        display_name=display_name,
        is_healthy=is_healthy,
        category=category,
    )
